apisix/plugins/sls-logger.lua

```diff
@@ -0,0 +1,206 @@
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local log_util = require("apisix.utils.log-util")
+local batch_processor = require("apisix.utils.batch-processor")
+local plugin_name = "sls-logger"
+local ngx = ngx
+local rf5424 = require("apisix.plugins.slslog.rfc5424")
+local stale_timer_running = false;
+local timer_at = ngx.timer.at
+local tcp = ngx.socket.tcp
+local buffers = {}
+local tostring = tostring
+local ipairs = ipairs
+local schema = {
+    type = "object",
+    properties = {
+        include_req_body = {type = "boolean", default = false},
+        name = {type = "string", default = "sls-logger"},
+        timeout = {type = "integer", minimum = 1, default= 5000},
+        max_retry_count = {type = "integer", minimum = 0, default = 0},
+        retry_delay = {type = "integer", minimum = 0, default = 1},
+        buffer_duration = {type = "integer", minimum = 1, default = 60},
+        inactive_timeout = {type = "integer", minimum = 1, default = 5},
+        batch_max_size = {type = "integer", minimum = 1, default = 1000},
+        host = {type = "string"},
+        port = {type = "integer"},
+        project = {type = "string"},
+        logstore = {type = "string"},
+        access_key_id = {type = "string"},
+        access_key_secret = {type ="string"}
+    },
+    required = {"host", "port", "project", "logstore", "access_key_id", "access_key_secret"}
+}
+
+local _M = {
+    version = 0.1,
+    priority = 406,
+    name = plugin_name,
+    schema = schema,
+}
+
+function _M.check_schema(conf)
+   return core.schema.check(schema, conf)
+end
+
+local function send_tcp_data(route_conf, log_message)
+    local err_msg
+    local res = true
+    local sock, soc_err = tcp()
+    local can_close
+
+    if not sock then
+        return false, "failed to init the socket" .. soc_err
+    end
+
+    sock:settimeout(route_conf.timeout)
+    local ok, err = sock:connect(route_conf.host, route_conf.port)
+    if not ok then
+        return false, "failed to connect to TCP server: host[" .. route_conf.host
+                      .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err
+    end
+
+    ok, err = sock:sslhandshake(true, nil, false)
+    if not ok then
+        return false, "failed to to perform TLS handshake to TCP server: host["
+                      .. route_conf.host .. "] port[" .. tostring(route_conf.port)
+                      .. "] err: " .. err
+    end
+
+    core.log.debug("sls logger send data ", log_message)
+    ok, err = sock:send(log_message)
+    if not ok then
+        res = false
+        can_close = true
+        err_msg = "failed to send data to TCP server: host[" .. route_conf.host
+                  .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err
+    else
+        ok, err = sock:setkeepalive(120 * 1000, 20)
+        if not ok then
+            can_close = true
+            core.log.warn("failed to set socket keepalive: host[" .. route_conf.host
```

Comment:
bad performance, avoid concat string in Lua land.

`core.log.warn("xxxx [", conf.host, "] port[", ...)`, please use this way. and please update other similar points

Reply:
fix it.